File: hdf-diff/test/apply-events.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { validateResults } from '@mitre/hdf-validators';
import { changeEventFromPrevious, type KeyState } from '../src/change-event.js';
import { computeEffectiveChecksum, computeEffectiveImpact } from '../src/effective-checksum.js';
import { computeEffectiveStatus } from '../src/status.js';
import { applyChangeEvents, type ApplyInputs } from '../src/apply-events.js';

const FIXTURES = join(dirname(fileURLToPath(import.meta.url)), 'fixtures');

function loadFixture(name: string): string {
  return readFileSync(join(FIXTURES, name), 'utf-8');
}

function applyInputs(): ApplyInputs {
  return {
    generator: { name: 'conmon-reconciler-test', version: '0.0.1' },
    seedUri: 'seed.hdf.json',
    source: 'inspec://fixture/scan',
  };
}

type Req = Record<string, unknown>;

function requirementsOf(doc: Record<string, unknown>): Req[] {
  const baselines = (doc['baselines'] as Record<string, unknown>[] | undefined) ?? [];
  return baselines.flatMap((b) => (b['requirements'] as Req[] | undefined) ?? []);
}

/** Derive the event stream between two same-target docs via the real kernel. */
async function deriveStream(seedJson: string, nextJson: string): Promise<Req[]> {
  const seedDoc = JSON.parse(seedJson) as Record<string, unknown>;
  const nextDoc = JSON.parse(nextJson) as Record<string, unknown>;
  const refTs = (nextDoc['timestamp'] as string | undefined) ?? '2026-07-22T14:03:11Z';

  const prevByKey = new Map<string, { state: KeyState; req: Req }>();
  for (const r of requirementsOf(seedDoc)) {
    prevByKey.set(r['id'] as string, {
      state: {
        effectiveStatus: computeEffectiveStatus(r, refTs),
        effectiveImpact: computeEffectiveImpact(r, refTs),
        checksum: await computeEffectiveChecksum(r, refTs),
      },
      req: r,
    });
  }

  let seq = 0;
  const mkInputs = (id: string) => ({
    eventId: `0190f6f2-0000-7000-8000-${String(++seq).padStart(12, '0')}`,
    source: 'inspec://fixture/scan',
    sequence: seq,
    systemRef: 'fixture.hdf-system.json',
    componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
    requirementId: id,
    timestamp: '2026-07-22T14:03:11Z',
    referenceTimestamp: refTs,
  });

  const events: Req[] = [];
  for (const r of requirementsOf(nextDoc)) {
    const id = r['id'] as string;
    const prev = prevByKey.get(id) ?? null;
    prevByKey.delete(id);
    const ev = await changeEventFromPrevious(prev?.state ?? null, r, prev?.req ?? null, mkInputs(id));
    if (ev) events.push(ev);
  }
  for (const id of [...prevByKey.keys()].sort()) {
    const prev = prevByKey.get(id)!;
    const ev = await changeEventFromPrevious(prev.state, null, prev.req, mkInputs(id));
    if (ev) events.push(ev);
  }
  return events;
}

function maskVolatile(req: Req): void {
  for (const r of (req['results'] as Req[] | undefined) ?? []) {
    delete r['startTime'];
    delete r['runTime'];
  }
}

async function parityCheck(seedName: string, nextName: string): Promise<void> {
  const seedJson = loadFixture(seedName);
  const nextJson = loadFixture(nextName);
  const events = await deriveStream(seedJson, nextJson);
  const changed = new Set(events.map((e) => e['requirementId'] as string));

  const { results, warnings } = await applyChangeEvents(seedJson, events, applyInputs());
  expect(warnings, 'a complete derived stream must verify cleanly').toEqual([]);

  const got = new Map(requirementsOf(results).map((r) => [r['id'] as string, r]));
  const want = new Map(
    requirementsOf(JSON.parse(nextJson) as Record<string, unknown>).map((r) => [r['id'] as string, r]),
  );
  expect(got.size).toBe(want.size);
  for (const [id, wantReq] of want) {
    const gotReq = got.get(id);
    expect(gotReq, `requirement ${id} missing from reassembled doc`).toBeDefined();
    const w = structuredClone(wantReq);
    const g = structuredClone(gotReq!);
    if (!changed.has(id)) {
      maskVolatile(w);
      maskVolatile(g);
    }
    expect(g, `requirement ${id} must match the rescan content`).toEqual(w);
  }

  const validation = validateResults(results);
  expect(validation.valid, JSON.stringify(validation.errors)).toBe(true);
}

describe('applyChangeEvents', () => {
  it('satisfies the reassembly-parity law on the scan pair', async () => {
    await parityCheck('scan-before.json', 'scan-after.json');
  });

  it('satisfies the reassembly-parity law on the override pair', async () => {
    await parityCheck('scan-before.json', 'scan-with-override.json');
  });

  it('is idempotent under duplicate delivery', async () => {
    const seedJson = loadFixture('scan-before.json');
    const events = await deriveStream(seedJson, loadFixture('scan-after.json'));
    const once = await applyChangeEvents(seedJson, events, applyInputs());
    const twice = await applyChangeEvents(seedJson, [...events, ...events], applyInputs());
    expect(twice.results).toEqual(once.results);
  });

  it('is invariant under delivery order', async () => {
    const seedJson = loadFixture('scan-before.json');
    const events = await deriveStream(seedJson, loadFixture('scan-after.json'));
    const inOrder = await applyChangeEvents(seedJson, events, applyInputs());
    const reversed = await applyChangeEvents(seedJson, [...events].reverse(), applyInputs());
    expect(reversed.results).toEqual(inOrder.results);
  });

  it('detects a chain gap but still applies last-value-wins', async () => {
    const seedJson = loadFixture('scan-before.json');
    const events = await deriveStream(seedJson, loadFixture('scan-after.json'));
    // Drop the fixed-transition event for SV-001, then send a later crafted
    // event for the same key whose prior links to the dropped posture.
    const sv1 = events.find((e) => e['requirementId'] === 'SV-001')!;
    const rest = events.filter((e) => e['requirementId'] !== 'SV-001');
    const after = sv1['after'] as Req;
    const followOn = await changeEventFromPrevious(
      {
        effectiveStatus: 'passed',
        effectiveImpact: 0.7,
        checksum: await computeEffectiveChecksum(after, '2026-07-22T14:03:11Z'),
      },
      { ...after, results: [{ status: 'failed', codeDesc: 'regressed again', startTime: '2026-07-23T00:00:00Z' }] },
      after,
      {
        eventId: '0190f6f2-0000-7000-8000-000000000099',
        source: 'inspec://fixture/scan',
        sequence: 99,
        systemRef: 'fixture.hdf-system.json',
        componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
        requirementId: 'SV-001',
        timestamp: '2026-07-23T00:00:00Z',
        referenceTimestamp: '2026-07-23T00:00:00Z',
      },
    );
    const { warnings } = await applyChangeEvents(seedJson, [...rest, followOn!], applyInputs());
    const gap = warnings.filter((w) => w.kind === 'chainGap');
    expect(gap).toHaveLength(1);
    expect(gap[0].requirementId).toBe('SV-001');
  });

  it('warns absentUnknown for a tombstone on an unknown key', async () => {
    const seedJson = loadFixture('scan-before.json');
    const ghost = await changeEventFromPrevious(
      {
        effectiveStatus: 'failed',
        effectiveImpact: 0.5,
        checksum: {
          algorithm: 'sha256',
          value: '704f62b2d0803438ad6b7b9bab45e2c4f350b7344135a2a7f8ef986d98669021',
        },
      },
      null,
      null,
      {
        eventId: '0190f6f2-0000-7000-8000-000000000777',
        source: 'inspec://fixture/scan',
        sequence: 7,
        systemRef: 'fixture.hdf-system.json',
        componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
        requirementId: 'SV-999999',
        timestamp: '2026-07-22T14:03:11Z',
        referenceTimestamp: '2026-07-22T14:03:11Z',
      },
    );
    const { warnings } = await applyChangeEvents(seedJson, [ghost!], applyInputs());
    expect(warnings).toHaveLength(1);
    expect(warnings[0].kind).toBe('absentUnknown');
  });

  it('passes the seed through with lineage on an empty batch', async () => {
    const seedJson = loadFixture('scan-before.json');
    const { results, warnings } = await applyChangeEvents(seedJson, [], applyInputs());
    expect(warnings).toEqual([]);
    const derivation = results['derivation'] as Req;
    expect(derivation['eventsApplied']).toBe(0);
    expect(derivation['throughSequence']).toBe(0);
    expect(typeof derivation['asOf']).toBe('string');
    const seedRef = derivation['seed'] as Req;
    expect(seedRef['uri']).toBe('seed.hdf.json');
    expect((seedRef['checksum'] as Req)['algorithm']).toBe('sha256');
    expect((results['generator'] as Req)['name']).toBe('conmon-reconciler-test');
  });
});

describe('applyChangeEvents defensive edges', () => {
  const MINI_SEED = JSON.stringify({
    timestamp: '2026-07-01T00:00:00Z',
    baselines: [
      { name: 'a', requirements: [{ id: 'R1', impact: 0.5, tags: {}, descriptions: [{ label: 'default', data: 'd' }], results: [{ status: 'failed', codeDesc: 't', startTime: '2025-01-01T00:00:00Z' }] }] },
      { name: 'b', requirements: [] },
    ],
  });

  function rawEvent(overrides: Record<string, unknown>): Record<string, unknown> {
    return {
      eventId: '0190f6f2-0000-7000-8000-000000000501',
      source: 'inspec://edge/scan',
      sequence: 501,
      systemRef: 'edge.hdf-system.json',
      componentId: '6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60',
      requirementId: 'R-EDGE',
      timestamp: '2026-07-02T00:00:00Z',
      priorChecksum: null,
      state: 'new',
      before: null,
      after: {
        id: 'R-EDGE', impact: 0.5, tags: {},
        descriptions: [{ label: 'default', data: 'd' }],
        results: [{ status: 'passed', codeDesc: 't', startTime: '2026-07-02T00:00:00Z' }],
      },
      ...overrides,
    };
  }

  it('warns multiBaseline when inserting a new key into a multi-baseline seed', async () => {
    const { warnings, results } = await applyChangeEvents(MINI_SEED, [rawEvent({})], applyInputs());
    expect(warnings.some((w) => w.kind === 'multiBaseline')).toBe(true);
    const first = (results['baselines'] as Req[])[0];
    expect((first['requirements'] as Req[]).some((r) => r['id'] === 'R-EDGE')).toBe(true);
  });

  it('throws when a new key arrives and the seed has no baselines', async () => {
    const seed = JSON.stringify({ timestamp: '2026-07-01T00:00:00Z', baselines: [] });
    await expect(applyChangeEvents(seed, [rawEvent({})], applyInputs())).rejects.toThrow(
      /no baselines/,
    );
  });

  it('throws when asOf is underivable (no events, no seed timestamp)', async () => {
    const seed = JSON.stringify({ baselines: [] });
    await expect(applyChangeEvents(seed, [], applyInputs())).rejects.toThrow(/cannot derive asOf/);
  });

  it('falls back to the seed timestamp for asOf when an event timestamp is unparseable', async () => {
    // A malformed event timestamp → parseTimestamp returns null → asOf is not
    // taken from the event; it falls back to the seed document's timestamp.
    const { results } = await applyChangeEvents(
      MINI_SEED,
      [rawEvent({ timestamp: 'not-a-date' })],
      applyInputs(),
    );
    const derivation = results['derivation'] as Record<string, unknown>;
    expect(derivation['asOf']).toBe('2026-07-01T00:00:00Z');
  });

  it('re-renders an offset-bearing event timestamp as trimmed UTC in asOf (Go parity)', async () => {
    // Go emits maxOccurred.UTC().Format(RFC3339Nano); a verbatim offset
    // string would diverge byte-wise between the two implementations.
    const { results } = await applyChangeEvents(
      MINI_SEED,
      [rawEvent({ timestamp: '2026-07-02T05:00:00+02:00' })],
      applyInputs(),
    );
    const derivation = results['derivation'] as Record<string, unknown>;
    expect(derivation['asOf']).toBe('2026-07-02T03:00:00Z');
    expect(results['timestamp']).toBe('2026-07-02T03:00:00Z');
  });

  it('warns chainGap on a content-bearing chain for a key the seed does not carry', async () => {
    const { warnings } = await applyChangeEvents(
      MINI_SEED,
      [rawEvent({ state: 'updated', before: { effectiveStatus: 'failed', effectiveImpact: 0.5 }, priorChecksum: { algorithm: 'sha256', value: 'deadbeef' } })],
      applyInputs(),
    );
    const gaps = warnings.filter((w) => w.kind === 'chainGap');
    expect(gaps).toHaveLength(1);
    expect(gaps[0].requirementId).toBe('R-EDGE');
    // The two-baseline seed also legitimately warns multiBaseline on insertion.
    expect(warnings.some((w) => w.kind === 'multiBaseline')).toBe(true);
  });

  it('warns chainGap when the winning event carries no after payload', async () => {
    const { warnings } = await applyChangeEvents(
      MINI_SEED,
      [rawEvent({ requirementId: 'R1', state: 'updated', after: null, before: { effectiveStatus: 'failed', effectiveImpact: 0.5 } })],
      applyInputs(),
    );
    expect(warnings.some((w) => w.kind === 'chainGap')).toBe(true);
  });
});

File: hdf-diff/go/apply_events_test.go
package diff

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"testing"
	"time"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	validators "github.com/mitre/hdf-libs/hdf-validators/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func aeInputs() ApplyInputs {
	return ApplyInputs{
		Generator: hdf.Generator{Name: "conmon-reconciler-test", Version: "0.0.1"},
		SeedURI:   "seed.hdf.json",
		Source:    "inspec://fixture/scan",
	}
}

// aeLoadFixture reads a shared fixture document.
func aeLoadFixture(t *testing.T, name string) []byte {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(fixturesDir(t), name))
	require.NoError(t, err)
	return data
}

// aeDeriveStream builds the event stream between two same-target documents
// using the real detection kernel — the derive() side of the §6 parity law.
// Sequences are assigned deterministically (new-doc order, then removed keys
// sorted by id).
func aeDeriveStream(t *testing.T, seedBytes, nextBytes []byte) []*hdf.HDFRequirementChangeEvent {
	t.Helper()
	var seedDoc, nextDoc hdf.HDFResults
	require.NoError(t, json.Unmarshal(seedBytes, &seedDoc))
	require.NoError(t, json.Unmarshal(nextBytes, &nextDoc))

	refTs := "2026-07-22T14:03:11Z"
	if nextDoc.Timestamp != nil {
		refTs = nextDoc.Timestamp.UTC().Format(time.RFC3339Nano)
	}
	prevRefTs := refTs
	if seedDoc.Timestamp != nil {
		prevRefTs = seedDoc.Timestamp.UTC().Format(time.RFC3339Nano)
	}

	type prevEntry struct {
		state KeyState
		req   hdf.EvaluatedRequirement
	}
	prevByKey := map[string]prevEntry{}
	for _, b := range seedDoc.Baselines {
		for _, r := range b.Requirements {
			cs := ComputeEffectiveChecksum(r, refTs)
			require.NotNil(t, cs)
			prevByKey[r.ID] = prevEntry{
				state: KeyState{
					EffectiveStatus: ComputeEffectiveStatus(r, refTs),
					EffectiveImpact: ComputeEffectiveImpact(r, refTs),
					Checksum:        *cs,
				},
				req: r,
			}
		}
	}

	// Events occur when the next document was observed — its scan time.
	occurred, err := time.Parse(time.RFC3339, refTs)
	require.NoError(t, err)
	var events []*hdf.HDFRequirementChangeEvent
	seq := int64(0)
	mkInputs := func(id string) EventInputs {
		seq++
		return EventInputs{
			EventID:                fmt.Sprintf("0190f6f2-0000-7000-8000-%012d", seq),
			Source:                 "inspec://fixture/scan",
			Sequence:               seq,
			SystemRef:              "fixture.hdf-system.json",
			ComponentID:            "6e0f2a3b-9c01-4d5e-8f7a-1b2c3d4e5f60",
			RequirementID:          id,
			Timestamp:              occurred,
			ReferenceTimestamp:     refTs,
			PrevReferenceTimestamp: prevRefTs,
		}
	}

	for _, b := range nextDoc.Baselines {
		for i := range b.Requirements {
			r := b.Requirements[i]
			var prevState *KeyState
			var prevReq *hdf.EvaluatedRequirement
			if entry, ok := prevByKey[r.ID]; ok {
				s, q := entry.state, entry.req
				prevState, prevReq = &s, &q
				delete(prevByKey, r.ID)
			}
			if ev := ChangeEventFromPrevious(prevState, &r, prevReq, mkInputs(r.ID)); ev != nil {
				events = append(events, ev)
			}
		}
	}
	removed := make([]string, 0, len(prevByKey))
	for id := range prevByKey {
		removed = append(removed, id)
	}
	sort.Strings(removed)
	for _, id := range removed {
		entry := prevByKey[id]
		s, q := entry.state, entry.req
		if ev := ChangeEventFromPrevious(&s, nil, &q, mkInputs(id)); ev != nil {
			events = append(events, ev)
		}
	}
	return events
}

// aeRequirementsByID extracts every requirement from a results document,
// keyed by id, canonically marshaled for comparison.
func aeRequirementsByID(t *testing.T, docBytes []byte) map[string]string {
	t.Helper()
	var doc map[string]interface{}
	require.NoError(t, json.Unmarshal(docBytes, &doc))
	out := map[string]string{}
	baselines, _ := doc["baselines"].([]interface{})
	for _, bRaw := range baselines {
		b, _ := bRaw.(map[string]interface{})
		reqs, _ := b["requirements"].([]interface{})
		for _, rRaw := range reqs {
			r, _ := rRaw.(map[string]interface{})
			id, _ := r["id"].(string)
			raw, err := json.Marshal(r)
			require.NoError(t, err)
			out[id] = string(raw)
		}
	}
	return out
}

// aeMaskVolatile applies the §6 documented mask to one requirement map:
// per-run volatile result fields (startTime, runTime) — legitimate residue on
// UNCHANGED requirements, which honestly carry the seed's last observation.
func aeMaskVolatile(req map[string]interface{}) {
	results, _ := req["results"].([]interface{})
	for _, rRaw := range results {
		if r, ok := rRaw.(map[string]interface{}); ok {
			delete(r, "startTime")
			delete(r, "runTime")
		}
	}
}

// aeNormalize canonicalizes a requirement map for comparison: empty optional
// arrays at the requirement level are equivalent to the field being absent
// (the Go kernel's typed after-payload canonicalizes them away via
// omitempty; the schema treats both identically).
func aeNormalize(req map[string]interface{}) {
	for k, v := range req {
		if arr, ok := v.([]interface{}); ok && len(arr) == 0 && k != "descriptions" && k != "results" {
			delete(req, k)
		}
	}
}

func aeParityCheck(t *testing.T, seedName, nextName string) {
	t.Helper()
	seed := aeLoadFixture(t, seedName)
	next := aeLoadFixture(t, nextName)
	events := aeDeriveStream(t, seed, next)

	changed := map[string]bool{}
	for _, ev := range events {
		changed[ev.RequirementID] = true
	}

	result, err := ApplyChangeEvents(seed, events, aeInputs())
	require.NoError(t, err)
	assert.Empty(t, result.Warnings, "a complete derived stream must verify cleanly")

	got := aeRequirementsByID(t, result.Results)
	want := aeRequirementsByID(t, next)
	require.Equal(t, len(want), len(got), "requirement count must match the rescan")
	for id, wantRaw := range want {
		gotRaw, ok := got[id]
		require.True(t, ok, "requirement %s missing from reassembled doc", id)

		var wantReq, gotReq map[string]interface{}
		require.NoError(t, json.Unmarshal([]byte(wantRaw), &wantReq))
		require.NoError(t, json.Unmarshal([]byte(gotRaw), &gotReq))
		aeNormalize(wantReq)
		aeNormalize(gotReq)
		if !changed[id] {
			// Unchanged keys: the reconciled doc honestly reports the last
			// observation; only per-run volatile fields may differ (§6 mask).
			aeMaskVolatile(wantReq)
			aeMaskVolatile(gotReq)
		}
		assert.Equal(t, wantReq, gotReq, "requirement %s must match the rescan content", id)
	}

	validation := validators.ValidateResults(result.Results)
	assert.True(t, validation.Valid, "reassembled doc must be schema-valid: %v", validation.Errors)
}

func TestApplyChangeEvents_ReassemblyParity_ScanPair(t *testing.T) {
	aeParityCheck(t, "scan-before.json", "scan-after.json")
}

func TestApplyChangeEvents_ReassemblyParity_OverridePair(t *testing.T) {
	// Card AC named the ubuntu pair, but those fixtures are legacy-shaped
	// (no baselines[].requirements) and unusable for a v3 reassembly law.
	// scan-before → scan-with-override exercises the override-driven change
	// path plus heavy tombstoning (four keys leave scope).
	aeParityCheck(t, "scan-before.json", "scan-with-override.json")
}

func TestApplyChangeEvents_DerivationBlockAndGenerator(t *testing.T) {
	seed := aeLoadFixture(t, "scan-before.json")
	next := aeLoadFixture(t, "scan-after.json")
	events := aeDeriveStream(t, seed, next)

	result, err := ApplyChangeEvents(seed, events, aeInputs())
	require.NoError(t, err)

	var doc map[string]interface{}
	require.NoError(t, json.Unmarshal(result.Results, &doc))

	gen, ok := doc["generator"].(map[string]interface{})
	require.True(t, ok, "reconciled doc must carry the reconciler generator")
	assert.Equal(t, "conmon-reconciler-test", gen["name"])

	deriv, ok := doc["derivation"].(map[string]interface{})
	require.True(t, ok, "reconciled doc must carry the derivation block")
	seedRef := deriv["seed"].(map[string]interface{})
	assert.Equal(t, "seed.hdf.json", seedRef["uri"])
	cs := seedRef["checksum"].(map[string]interface{})
	assert.Equal(t, "sha256", cs["algorithm"])
	assert.Len(t, cs["value"], 64)
	assert.Equal(t, float64(len(events)), deriv["eventsApplied"])
	var maxSeq int64
	for _, ev := range events {
		if ev.Sequence > maxSeq {
			maxSeq = ev.Sequence
		}
	}
	assert.Equal(t, float64(maxSeq), deriv["throughSequence"],
		"watermark is the highest sequence seen (unchanged keys still consume sequence numbers)")
	assert.Equal(t, "inspec://fixture/scan", deriv["source"])
	assert.Equal(t, "2024-02-01T00:00:00Z", deriv["asOf"], "asOf is the next scan's observation time")
}

func TestApplyChangeEvents_Idempotent(t *testing.T) {
	seed := aeLoadFixture(t, "scan-before.json")
	next := aeLoadFixture(t, "scan-after.json")
	events := aeDeriveStream(t, seed, next)

	once, err := ApplyChangeEvents(seed, events, aeInputs())
	require.NoError(t, err)
	twice, err := ApplyChangeEvents(seed, append(append([]*hdf.HDFRequirementChangeEvent{}, events...), events...), aeInputs())
	require.NoError(t, err)
	assert.Equal(t, string(once.Results), string(twice.Results),
		"duplicate delivery (same source+eventId) must not change the output")
}

func TestApplyChangeEvents_OrderInvariant(t *testing.T) {
	seed := aeLoadFixture(t, "scan-before.json")
	next := aeLoadFixture(t, "scan-after.json")
	events := aeDeriveStream(t, seed, next)

	reversed := make([]*hdf.HDFRequirementChangeEvent, len(events))
	for i, ev := range events {
		reversed[len(events)-1-i] = ev
	}

	inOrder, err := ApplyChangeEvents(seed, events, aeInputs())
	require.NoError(t, err)
	shuffled, err := ApplyChangeEvents(seed, reversed, aeInputs())
	require.NoError(t, err)
	assert.Equal(t, string(inOrder.Results), string(shuffled.Results),
		"delivery order must not affect the fold (sequence is the only ordering authority)")
}

// aeTwoStepChain builds seed bytes plus a two-event chain for one key:
// failed → passed (fixed, seq 1) → failed (regressed, seq 2), with event 2's
// priorChecksum linking to event 1's after-posture.
func aeTwoStepChain(t *testing.T) (seed []byte, e1, e2 *hdf.HDFRequirementChangeEvent) {
	t.Helper()
	seedJSON := `{
	  "timestamp": "2026-07-22T00:00:00Z",
	  "baselines": [{"name": "chain", "requirements": [
	    {"id": "SV-100001", "impact": 0.5, "tags": {},
	     "descriptions": [{"label": "default", "data": "test description"}],
	     "results": [{"status": "failed", "codeDesc": "test", "startTime": "2025-01-01T00:00:00Z"}]}
	  ]}],
	  "statistics": {"duration": 0.1}
	}`
	refTs := "2026-07-22T14:03:11Z"

	seedReq := ceValid(ecFailingReq())
	seedReq.ID = "SV-100001"
	seedCS := ComputeEffectiveChecksum(seedReq, refTs)
	require.NotNil(t, seedCS)

	passing := cePassingReq()
	passing.ID = "SV-100001"
	in1 := ceInputs()
	in1.RequirementID = "SV-100001"
	in1.Sequence = 1
	in1.EventID = "0190f6f2-0000-7000-8000-000000000001"
	in1.ReferenceTimestamp = refTs
	e1 = ChangeEventFromPrevious(&KeyState{
		EffectiveStatus: "failed", EffectiveImpact: 0.5, Checksum: *seedCS,
	}, &passing, &seedReq, in1)
	require.NotNil(t, e1)

	afterCS := ComputeEffectiveChecksum(passing, refTs)
	require.NotNil(t, afterCS)
	failingAgain := ceValid(ecFailingReq())
	failingAgain.ID = "SV-100001"
	in2 := in1
	in2.Sequence = 2
	in2.EventID = "0190f6f2-0000-7000-8000-000000000002"
	e2 = ChangeEventFromPrevious(&KeyState{
		EffectiveStatus: "passed", EffectiveImpact: 0.5, Checksum: *afterCS,
	}, &failingAgain, &passing, in2)
	require.NotNil(t, e2)

	return []byte(seedJSON), e1, e2
}

func TestApplyChangeEvents_ChainVerification(t *testing.T) {
	seed, e1, e2 := aeTwoStepChain(t)

	t.Run("complete chain verifies cleanly", func(t *testing.T) {
		result, err := ApplyChangeEvents(seed, []*hdf.HDFRequirementChangeEvent{e1, e2}, aeInputs())
		require.NoError(t, err)
		assert.Empty(t, result.Warnings)
	})

	t.Run("missing link is detected but still applied last-value-wins", func(t *testing.T) {
		full, err := ApplyChangeEvents(seed, []*hdf.HDFRequirementChangeEvent{e1, e2}, aeInputs())
		require.NoError(t, err)
		gapped, err := ApplyChangeEvents(seed, []*hdf.HDFRequirementChangeEvent{e2}, aeInputs())
		require.NoError(t, err)

		require.Len(t, gapped.Warnings, 1)
		assert.Equal(t, "chainGap", gapped.Warnings[0].Kind)
		assert.Equal(t, "SV-100001", gapped.Warnings[0].RequirementID)

		// The winning posture is identical either way — only the derivation
		// metadata (eventsApplied) may differ.
		fullReqs := aeRequirementsByID(t, full.Results)
		gappedReqs := aeRequirementsByID(t, gapped.Results)
		assert.Equal(t, fullReqs, gappedReqs)
	})
}

func TestApplyChangeEvents_Anomalies(t *testing.T) {
	seed, e1, _ := aeTwoStepChain(t)

	t.Run("absent for an unknown key warns and no-ops", func(t *testing.T) {
		in := ceInputs()
		in.RequirementID = "SV-999999"
		in.Sequence = 7
		ghost := ChangeEventFromPrevious(&KeyState{
			EffectiveStatus: "failed", EffectiveImpact: 0.5,
			Checksum: hdf.Checksum{Algorithm: hdf.Sha256, Value: ecVectorFailedHalf},
		}, nil, nil, in)
		require.NotNil(t, ghost)

		result, err := ApplyChangeEvents(seed, []*hdf.HDFRequirementChangeEvent{ghost}, aeInputs())
		require.NoError(t, err)
		require.Len(t, result.Warnings, 1)
		assert.Equal(t, "absentUnknown", result.Warnings[0].Kind)
		assert.Len(t, aeRequirementsByID(t, result.Results), 1, "seed requirement untouched")
	})

	t.Run("new for an existing key warns and replaces", func(t *testing.T) {
		passing := cePassingReq()
		passing.ID = "SV-100001"
		in := ceInputs()
		in.RequirementID = "SV-100001"
		in.Sequence = 9
		fresh := ChangeEventFromPrevious(nil, &passing, nil, in)
		require.NotNil(t, fresh)

		result, err := ApplyChangeEvents(seed, []*hdf.HDFRequirementChangeEvent{fresh}, aeInputs())
		require.NoError(t, err)
		require.Len(t, result.Warnings, 1)
		assert.Equal(t, "newOnExisting", result.Warnings[0].Kind)
	})

	t.Run("empty event batch is a seed passthrough with lineage", func(t *testing.T) {
		result, err := ApplyChangeEvents(seed, nil, aeInputs())
		require.NoError(t, err)
		assert.Empty(t, result.Warnings)
		var doc map[string]interface{}
		require.NoError(t, json.Unmarshal(result.Results, &doc))
		deriv := doc["derivation"].(map[string]interface{})
		assert.Equal(t, float64(0), deriv["eventsApplied"])
		assert.Equal(t, "2026-07-22T00:00:00Z", deriv["asOf"], "falls back to the seed timestamp")
	})

	_ = e1
}
